small updates to Telemetry

robuye opened the pull request:
Hi there, I added some telemetry around webhook handler and changed metadata included in all API requests. It's all just a few small changes to make monitoring for stripity-stripe users awesome out of the box 🙃 

Webhooks are often important part of integration with Stripe so having telemetry in there is good idea imo.

API requests metadata contains some low cardinality fields that can be useful when visualising these metrics in other tools. For example using `stripe.request.stop` event we can infer metrics for RED method and also have visibility into individual Stripe API endpoints.

There are no breaking changes in here, only telemetry metadata is different (`url` field is gone). There is assumption that `event` object always contains `type` property, which is always a case to my knowledge and Stripe docs.

Here is example how it looks in PhoenixLiveDashboard:

![image](https://github.com/beam-community/stripity-stripe/assets/853568/1d286b8e-2d04-4435-83bc-426158802ff2)

In order to enable this visualisation I only needed to add few lines to `telemetry.ex`:

```elixir
      # Stripe
      summary("stripe.request.stop.duration",
        tags: [:endpoint, :status],
        unit: {:native, :millisecond}
      ),
      summary("stripe.webhook.stop.duration",
        tags: [:event],
        unit: {:native, :millisecond}
      ),
```

Other systems such as prom_ex, telemetry_ui, otel, etc can be easily integrated too.

Hope you like it, happy to take in any and all feedback.

robuye committed:
feat: add metadata to telemetry events

Added telemetry support to webhook handler and changed metadata included
in API requests.

API requests metadata contains some low cardinality fields that can be
useful when visualising these metrics in other tools. For example using
'stripe.request.stop' event we can infer metrics for RED method and have
visibility into all Stripe API endpoints.

'url' passed in to 'do_perform_request_and_retry' contains query params
which can contain sensitive data such as emails so it is URI-parsed and
all params are dropped.

Webhooks are often important part of integration with Stripe and having
telemetry in place for that is good idea too imo.

yordis commented:
CI is failing btw

robuye committed:
chore: configure SHELL env in CI jobs

Maybe fix problem with erlexec failing to start:

```
SHELL environment variable not set! [c_src/exec.cpp:609]

Notice: 15:48:19.772 [notice] Application erlexec exited: :exec_app.start(:normal, []) returned an error: shutdown: failed to start child: :exec
    ** (EXIT) {:port_exited_with_status, 4}
```

robuye committed:
chore: cleanup CI output

Run mix compile before tests so the output is separate.

Trace will make it easier to understand if some tests were excluded and
such.

robuye committed:
chore: fix flaky tests

Some tests are failing because query params are in wrong order.

My fix is to ignore order and just checking all were set with correct
value.

robuye committed:
chore: run mix format and add it to CI

robuye commented:
Okay, all tests are passing now, I think I fixed original problem by setting SHELL env. I also noticed `mix format` is updating 2 files so I fixed that too and added format checking to CI job.

yordis commented on lib/stripe/webhook_plug.ex:
```diff
@@ -159,30 +159,38 @@ if Code.ensure_loaded?(Plug) do
-          Event data: #{inspect(event)}
-          """
-      end
+      :telemetry.span(~w[stripe webhook]a, %{}, fn ->
```
Couldn't this be done at the WebServer Telemetry level? That was the feedback given from the OTEL folks

robuye replied:
I think yes, Phoenix and Plug already add their own events around requests, but these are very generic and we can do it better here by annotating with Stripe event type. Having this together with other part gives us more complete view of Stripe subsystem, or that's how I see it from monitoring perspective.

yordis replied:
I am with you! I see the point of adding `event.type` since it is domain-specific.

But wouldn't the WebServer handle the `handler_status` based on the response?

robuye replied:
btw I'm actually working on integrating OTEL and added Telemetry here to create spans and traces similar to how it's done in phoenix in [otel-erlang-contrib](https://github.com/open-telemetry/opentelemetry-erlang-contrib/blob/main/instrumentation/opentelemetry_phoenix/lib/opentelemetry_phoenix.ex#L317-L344). I hope to have spans that include stripe api calls by name, I think that will be super helpful when browsing traces, but would be difficult to do using generic telemetry events from phoenix or other lib.

yordis replied:
I am not opposed to anything, but since these are architectural components, we are better off doing research and making sure we are following "standards" (OTEL practices for now) so we all benefit

I came across particular telemetry requirements that required the Operator to change their mindset and workflow to accomplish what they wanted; I am not saying that is the case here! I am just asking questions to make sure we are doing the right thing. I hope you understand.

This is a tricky domain; everyone wants their way of observing things...

robuye replied:
> But wouldn't the WebServer handle the handler_status based on the response?

Well, sort of. WebServer will report it in its own telemetry as completely different event so in a sense yes, but it's not always available. I think OTEL traces will be able to connect to parent span and present this info but other systems, for example Prometheus, won't.

Adding handler_status to this event makes it possible to monitor error rates by Stripe event type and use RED method to get useful insights with little effort, that's the main reason why I decided to add it here.

Alternatively we could add exactly same tracing to handler module that `WebhookHandler` is calling. It will work the same but this is additional effort to build custom observability versus having it provided ootb by stripity-stripe.

I personally prefer ootb experience because otherwise it's difficult to get in and do it right on your own. 

> I am not opposed to anything, but since these are architectural components, we are better off doing research and making sure we are following "standards" (OTEL practices for now) so we all benefit
> [...]

Yup, understood, happy to have this conversation and to look for best solution together, thanks for taking the time to discuss it too!

yordis commented on lib/stripe/api.ex:
```diff
@@ -377,13 +377,21 @@ defmodule Stripe.API do
+          status: nil
+        }
+
         case http_module().request(method, url, Map.to_list(headers), body, opts) do
```
```suggestion
        case http_module().request(method, url, Map.to_list(headers), body, opts) do
```

Notice the usage of `http_module()` which means this level of abstraction should be only about Stripe specific info (such as the header?!) but if people use `Req` or other clients, such `http_module()` may include its own Telemetry component.

robuye replied:
Telemetry should work fine with other clients, it's opt in and people need to configure their own handlers in code to use these events so there is no risk of conflict with other libs having their own telemetry. Metrics will ofc overlap to some extent, but our events have additional data that makes them meaningful in Stripe sense, such as API endpoint and version or event kind for webhooks. 

I added Stripe version header because it can uniquely identify API endpoint, dunno if anyone would miss it if it's not here. Actually status and request path are most useful, I can remove all others if you like.

You could get exactly same data by creating your own http_module and webhook handler and adding telemetry in there, but having it available out of the box and without need for custom adapters or other code is nice to have and has value imo.

yordis replied:
My point is that we could add the telemetry and lower its location, closer to the HTTP client itself; in the case of `hackney,` we need to provide the telemetry implementation; other clients like `Req` could use whatever available built-ins or add the layer to it.

What that means in practice is to change the following function to use our hackney module:

```diff
  @spec http_module() :: module
  def http_module() do
-    Config.resolve(:http_module, :hackney)
+    Config.resolve(:http_module, Stripe.Hackney) # <----- add the callback and telemetry
  end
```

I found a package that adds `:telemetry` to hackney already: https://github.com/msramos/hackney_telemetry

robuye replied:
Hmm, so the way it is now you can say that it adds layer to whatever http client provides. Users will be able to use our or their http client telemetry or both at the same time, we don't need to provide per client implementation.

If we don't add telemetry then users need to filter generic events for stripe requests probably based on path or hostname if these attributes are available. I bet many will be miserable doing that 🙃 😆 

If we add our telemetry then they can use our events and don't need to think about filtering and what events look like in their http client library, it will take just few lines of code to build dashboards like I shown in PR description. I could also can add a section to readme make our telemetry easier to use.

I think pushing code one level down makes sense if the goal is to clean up API module and push there more http concerns, for example retries, but functionally telemetry would continue to work the same as now, that is will be available regardless of telemetry provided by actual http client.

yordis commented on lib/stripe/api.ex:
```diff
@@ -376,14 +376,22 @@ defmodule Stripe.API do
 
   defp do_perform_request_and_retry(method, url, headers, body, opts, {:attempts, attempts}) do
+    telemetry_meta = %{
+      endpoint: URI.parse(url).path,
```
```suggestion
      http_url: URI.parse(url),
```

Pass the entire URL, let processors to filter the info, HTTP OTEL packages would do something around the line of https://github.com/open-telemetry/opentelemetry-erlang-contrib/blob/3f13bc8b2042071b723d0ea8c7b118fd10f42d26/instrumentation/opentelemetry_req/lib/opentelemetry_req.ex#L143-L148

robuye replied:
okay added and I renamed original field to `stripe_api_endpoint`. I can remove it because you're right that it can be extracted from `http_url` via processor, I kept it for better out-ot-the-box experience.

yordis commented on lib/stripe/api.ex:
```diff
@@ -376,14 +376,22 @@ defmodule Stripe.API do
   defp do_perform_request_and_retry(method, url, headers, body, opts, {:attempts, attempts}) do
+    telemetry_meta = %{
+      endpoint: URI.parse(url).path,
+      method: method,
```
```suggestion
      http_method: method,
```

yordis commented on lib/stripe/api.ex:
```diff
@@ -376,14 +376,22 @@ defmodule Stripe.API do
+      endpoint: URI.parse(url).path,
+      method: method,
+      attempt: attempts,
+      stripe_api_version: headers["Stripe-Version"],
```
```suggestion
      http_headers: headers,
```

yordis commented on lib/stripe/api.ex:
```diff
@@ -376,14 +376,22 @@ defmodule Stripe.API do
+      method: method,
+      attempt: attempts,
+      stripe_api_version: headers["Stripe-Version"],
+      status: nil
```
```suggestion
      http_status_code: nil
```

yordis commented on lib/stripe/api.ex:
```diff
@@ -376,14 +376,22 @@ defmodule Stripe.API do
+    telemetry_meta = %{
+      endpoint: URI.parse(url).path,
+      method: method,
+      attempt: attempts,
```
```suggestion
      http_retry_count: attempts,
```

yordis commented on lib/stripe/api.ex:
```diff
@@ -376,14 +376,22 @@ defmodule Stripe.API do
         case http_module().request(method, url, Map.to_list(headers), body, opts) do
           {:ok, status, _, _} = resp ->
-            {resp, %{status: status}}
+            {resp, %{telemetry_meta | status: status}}
```
```suggestion
            {resp, %{telemetry_meta | http_status_code: status, http_response_content_length: http_response_content_length }}
```

If `content-length` is present add `http_response_content_length`

yordis reviewed (changes requested):
Hey, just few minors adjustment, primarily around HTTP metadata and aligned a bit more with common OTEL lingos.

robuye committed:
fixup: update telemetry metadata

Updated naming and changed where metadata is constructed in order to
keep original function short and focused.

URL is stripped of query params to avoid leaking of sensitive data such
as emails to logging systems.

robuye commented:
Hii, I updated names per your feedback, but not completely.

1) I didn't include all headers because they can contain sensitive information such as cookies or api tokens
2) I strip query params from url so things like email don't end up in logs
3) I didn't include response content length because it requires parsing :hackney headers using low level API, this would make code more coupled to hackney adapter and it didn't feel like something you'd want

I also moved setting metadata one level higher so the original function is less complicated. Just experimenting and trying to fit in with current code patterns.

Lemme know what you think, if it looks good this way I will squash this commit with the original.

yordis commented on lib/stripe/api.ex:
```diff
@@ -217,6 +217,22 @@ defmodule Stripe.API do
+
+    opts
+    |> Keyword.put(:telemetry_metadata, %{
+      stripe_api_endpoint: parsed_uri.path,
```
This could be extracted from `http_url` if somebody really wants to map the data to something else

robuye commented:
Okay, I think I addressed earlier feedback in code or in GH and also left some followup questions for you, I'm not sure you noticed it 😅

doomspork commented:
Incorporated these commits and feedback in #895
